Add prometheus metrics for dockerregistry

legionus opened the pull request:
Example:
```

# HELP http_request_duration_microseconds The HTTP request latencies in microseconds.
# TYPE http_request_duration_microseconds summary
http_request_duration_microseconds{handler="openshift",quantile="0.5"} 12060.892
http_request_duration_microseconds{handler="openshift",quantile="0.9"} 26831.134
http_request_duration_microseconds{handler="openshift",quantile="0.99"} 92603.647
http_request_duration_microseconds_sum{handler="openshift"} 817404.5250000001
http_request_duration_microseconds_count{handler="openshift"} 52

# HELP http_request_size_bytes The HTTP request sizes in bytes.
# TYPE http_request_size_bytes summary
http_request_size_bytes{handler="openshift",quantile="0.5"} 345
http_request_size_bytes{handler="openshift",quantile="0.9"} 583
http_request_size_bytes{handler="openshift",quantile="0.99"} 3114
http_request_size_bytes_sum{handler="openshift"} 25907
http_request_size_bytes_count{handler="openshift"} 52

# HELP http_response_size_bytes The HTTP response sizes in bytes.
# TYPE http_response_size_bytes summary
http_response_size_bytes{handler="openshift",quantile="0.5"} 87
http_response_size_bytes{handler="openshift",quantile="0.9"} 2739
http_response_size_bytes{handler="openshift",quantile="0.99"} 2740
http_response_size_bytes_sum{handler="openshift"} 25128
http_response_size_bytes_count{handler="openshift"} 52

# HELP http_requests_total Total number of HTTP requests made.
# TYPE http_requests_total counter
http_requests_total{code="200",handler="openshift",method="get"} 19
http_requests_total{code="200",handler="openshift",method="head"} 6
http_requests_total{code="201",handler="openshift",method="post"} 1
http_requests_total{code="201",handler="openshift",method="put"} 4
http_requests_total{code="202",handler="openshift",method="patch"} 2
http_requests_total{code="202",handler="openshift",method="post"} 2
http_requests_total{code="400",handler="openshift",method="put"} 2
http_requests_total{code="401",handler="openshift",method="get"} 14
http_requests_total{code="404",handler="openshift",method="head"} 2

# HELP openshift_registry_request_duration_microseconds Request latency summary in microseconds for each operation
# TYPE openshift_registry_request_duration_microseconds summary
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.serveblob",quantile="0.5"} 316
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.serveblob",quantile="0.9"} 385
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.serveblob",quantile="0.99"} 385
openshift_registry_request_duration_microseconds_sum{name="tmp/busybox",operation="blobstore.serveblob"} 1180
openshift_registry_request_duration_microseconds_count{name="tmp/busybox",operation="blobstore.serveblob"} 3
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.stat",quantile="0.5"} 466
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.stat",quantile="0.9"} 632
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="blobstore.stat",quantile="0.99"} 632
openshift_registry_request_duration_microseconds_sum{name="tmp/busybox",operation="blobstore.stat"} 6633
openshift_registry_request_duration_microseconds_count{name="tmp/busybox",operation="blobstore.stat"} 7
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="manifestservice.get",quantile="0.5"} 10427
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="manifestservice.get",quantile="0.9"} 10427
openshift_registry_request_duration_microseconds{name="tmp/busybox",operation="manifestservice.get",quantile="0.99"} 10427
openshift_registry_request_duration_microseconds_sum{name="tmp/busybox",operation="manifestservice.get"} 22958
openshift_registry_request_duration_microseconds_count{name="tmp/busybox",operation="manifestservice.get"} 2
```

This is only one part of implemented metrics. Currently collected metrics about the process state, memstats, the garbage collector metrics.

legionus commented:
@mfojtik @miminar @smarterclayton PTAL

mfojtik commented:
@legionus I assume the metrics will only be visible for the cluster admin user, right?

@jcantrill FYI (this might be interesting for metrics)

legionus commented:
> I assume the metrics will only be visible for the cluster admin user, right?

@mfojtik No. Right now `/extensions/v2/metrics` open for anonymous access. But we can restrict it.

mfojtik commented:
@legionus you probably don't want to expose these data to anonymous users. also if we are going to include per-repo metrics (num of pulls), you want to make that info available only for admin (or the user that has access to the repo?).

legionus commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[registry] Registry Metrics](https://trello.com/c/y1pyZHl6/634-registry-registry-download-metrics)

deads2k commented:
Metrics about response time and registry performance should not be exposed to end users. They can and should be restricted to cluster-admins or another privileged ops kind of role.

Where do you envision per-repository metrics going, what would you expose (and how), and how do you plan to restrict access?

legionus commented:
@mfojtik @deads2k Sure. After discussion with @enj we planning to make some virtual object (`registrymetrics` for example) and use SAR to check access to it. I don't want to give any access in addition to receiving metrics.

legionus commented:
@deads2k I know it's ugly, but I don't know other way how to control access without giving any access to other resources.

jcantrill commented:
cc @mwringe

smarterclayton commented:
The approach established for app metrics is probably what we should follow - something associated with the registry container definition that is a shared secret that can be accessed by the metrics endpoint.

smarterclayton commented:
Prometheus recommends being careful on high cardinality metrics.  A good read is here: https://prometheus.io/docs/practices/instrumentation/

Will look more

smarterclayton commented:
Some metrics I would expect to be able to tell end users about:

1. How many times has anyone downloaded this image (we need to correlate that to one of the operations trivially)
2. How long did it take to download this image
3. How many requests to openshift API did we make over this period
4. Which openshift API calls took the most time to complete for the registry
5. Do we have high latency between the registry and etcd
6. When downloading blobs, what was the aggregate rate
7. How many bytes of blob data was served per unit time

mfojtik commented:
So we agreed we are not going to expose any data to end-users, but this endpoint will be for operator kind of users.
We have to talk to operations team what kind of metrics they would like to have and pick ones that are cheap to provide.
Then talk to @mwringe about advertising the prometheus endpoint to hawkular agent (pod annotations?).

The auth mechanism will be shared secret I assume, so think endpoint won't use the auth middleware.

dmage commented on pkg/dockerregistry/server/metrics/metrics.go:
```diff
@@ -0,0 +1,57 @@
+var registerMetrics sync.Once
+
+// Register all metrics.
+func Register() {
```
There is no reason to make this function public nor to use sync.Once: it's called from `init()` and should not be called manually.

smarterclayton replied:
We actually don't want to call things from init() unless we have to.  That registers metrics globally and pollutes the namespace.  We should only log metrics when we're going to use them.

legionus replied:
@smarterclayton Not a problem. I call `Register()` from `pkg/cmd/dockerregistry.Execute()` if metrics enabled.

legionus commented:
@mfojtik @miminar @smarterclayton I added an additional section in the config.yml because this functionality is not specific to any particular middleware. I added the option to enable metrics collection and the option to specify a shared secret.
It looks better for you ?

miminar commented on pkg/cmd/dockerregistry/dockerregistry.go:
```diff
@@ -135,9 +143,14 @@ func Execute(configFile io.Reader) {
-		context.GetLogger(app).Infof("listening on %v", config.HTTP.Addr)
-		if err := http.ListenAndServe(config.HTTP.Addr, handler); err != nil {
+	if extraConfig.Metrics.Enabled {
+		log.Debugf("configured metrics endpoint at \"/extensions/v2/metrics\"")
```
context.GetLogger(app)

miminar commented on pkg/cmd/dockerregistry/dockerregistry.go:
```diff
@@ -42,19 +42,22 @@ import (
 
 	ctx := context.Background()
-	ctx, err = configureLogging(ctx, config)
+	ctx = server.WithConfiguration(ctx, extraConfig)
```
Why does it need to be set on the context? Can it be passed to `RegisterMetricHandler` instead?

miminar replied:
Resolved eye-to-eye. If we're going with the top-level `openshift` section, this is the probably the only sensible way of passing the configuration to the registry/repository middleware.

miminar replied:
Nevertheless, I'd still want to avoid using the configuration from the context wherever possible. Let's use `server.AccessControllerParams` to configure the auth handler and pass the other configuration options directly to any other object we initialize here.

legionus replied:
@miminar Please no. Otherwise we have to add that to the repository middleware options as well. It already contains many parameters. I don't want to adв internal options to repository middleware.
Please take a look at my new changes.

miminar replied:
> I don't want to adв internal options to repository middleware.

@legionus It's fine to use the config object in `repository` middleware since we don't have a better way to pass the config there. On the other hand, there's no sense to have auth module differentiate between parameters passed via configuration file, parameters loaded from environment variables and parameters created on the fly. Those are assumptions that may easily become obsolete. Moreover, it makes unit tests hard to follow.

I still think we should pre-process the config and set `server.AccessControllerParams` accordingly. And do similar for all the other wrappers but `repository`. I won't block the review because of this though.
@dmage What is your opinion?

dmage replied:
I think passing variables through a context is a code smell, like a having global variables. It's very convenient to write this way (you can get data from anywhere), but it really complicates the code.

But in this case I don't care, I want to try to refactor the `repository` struct, after which there can emerge a place to pass the config. But it's too scary to do that without the good test coverage.

miminar replied:
Let's keep it as is. We can further tune it in a follow-up fixing #13568. Let's focus on the prometheus here.

miminar commented on pkg/dockerregistry/server/auditblobstore.go:
```diff
@@ -8,86 +8,179 @@ import (
 type blobWriter struct {
 	distribution.BlobWriter
+
+	repo *repository
```
Aren't we avoiding passing the repository everywhere? This could be just the `Named`.

miminar commented on pkg/dockerregistry/server/auditblobstore.go:
```diff
@@ -8,86 +8,179 @@ import (
 	return err
 }
 
 type blobWriter struct {
```
*nit*: Could this be renamed to `auditBlobWriter`?

miminar commented on pkg/dockerregistry/server/auditmanifestservice.go:
```diff
@@ -6,40 +6,78 @@ import (
 // write it in the audit log.
 type auditManifestService struct {
 	manifests distribution.ManifestService
+	repo      *repository
```
Just the name please.

miminar commented on pkg/dockerregistry/server/audittagservice.go:
```diff
@@ -5,47 +5,94 @@ import (
 // write it in the audit log.
 type auditTagService struct {
 	tags distribution.TagService
+	repo *repository
```
repo name

miminar commented on pkg/dockerregistry/server/auth.go:
```diff
@@ -312,7 +312,7 @@ func (ac *AccessController) Authorized(ctx context.Context, accessRecords ...reg
 	}
 
 	// In case of docker login, hits endpoint /v2
```
It looks like the comment is outdated.

miminar commented on pkg/dockerregistry/server/auth.go:
```diff
@@ -535,3 +545,18 @@ func verifyPruneAccess(ctx context.Context, client client.SubjectAccessReviews)
 }
+
+func isMetricsBearerToken(ctx context.Context, token string) bool {
+	config := ConfigurationFrom(ctx)
```
Could `AccessControllerParams` be used instead of the context?

miminar commented on pkg/dockerregistry/server/metrichandler.go:
```diff
@@ -0,0 +1,31 @@
+			},
+		}
+	}
+	extensionsRouter := app.NewRoute().PathPrefix("/extensions/v2/").Subrouter()
```
Could you turn this into a constant and use it on all the other places as well?

miminar commented on pkg/dockerregistry/server/metrics/dispatcher.go:
```diff
@@ -0,0 +1,17 @@
+
+	"github.com/docker/distribution/registry/handlers"
+
+	gorillahandlers "github.com/gorilla/handlers"
```
These could be squashed.

miminar commented on pkg/dockerregistry/server/auditblobstore.go:
```diff
@@ -8,86 +8,179 @@ import (
-	audit.GetLogger(ctx).Log("BlobStore.Stat")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.stat", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
```
I'd rather have audit and metric wrappers separated.

miminar commented on images/dockerregistry/config.yml:
```diff
@@ -35,3 +35,7 @@ middleware:
         blobrepositorycachettl: 10m
   storage:
     - name: openshift
+openshift:
```
Upstream has a configuration section called [reporting](https://docs.docker.com/registry/configuration/#reporting). This could be probably nested there - similarly to the `auth`.

miminar replied:
Resolved eye to eye. @legionus convinced me that this is a preferred place for our current and future configuration. As long as it's versioned properly. In the future, we'd like to move `middleware` section here completely. The `openshift` section can co-exist with the `middleware` entries for now since they have separate areas of interest. The `middleware` section gets out of hand already so I'd prefer to do the configuration refactoring sooner rather than later. I'll open an issue to track this.

miminar commented on pkg/dockerregistry/server/auditblobstore.go:
```diff
@@ -8,86 +8,179 @@ import (
 	err := b.store.Delete(ctx, dgst)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.Delete")
+
+	if audit.LoggerExists(ctx) {
```
Two many ifs. I think separated `metricBlobStore` would really make it more pleasant to the eye.

legionus commented:
@miminar Most of your commens addressed.

miminar commented on pkg/dockerregistry/server/metrics/metrics.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func (m *metricTimer) Stop() {
+	m.collector.WithLabelValues(m.labels...).Observe(float64(time.Since(m.startTime) / time.Microsecond))
```
`WithLabelValues` can panic. Could `GetMetricWithLabelValues` be used instead?

legionus replied:
@miminar and what I gona do in this function if `err != nil` ?

This function is used as a `defer` and I have at this moment no other possibility as to panic.

miminar replied:
I was thinking about logging the error. If the probability of panicking is close to zero, this is ok.

miminar commented on pkg/dockerregistry/server/metrics/metrics.go:
```diff
@@ -0,0 +1,47 @@
+	)
+)
+
+func NewTimer(collector *prometheus.SummaryVec, labels []string) *metricTimer {
```
GoDoc

miminar commented on pkg/dockerregistry/server/metrics/metrics.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+var (
+	RegistryAPIRequests = prometheus.NewSummaryVec(
```
Instead of having another global variable which is always set even when not needed, can we initialize this in in `dockerregistry.Execute()` and pass it in context as well?

miminar commented on pkg/dockerregistry/server/metrics/metricsmanifestservice.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/docker/distribution/digest"
+)
+
+// ManifestService wraps a distribution.ManifestService to collect statistic
```
statistics

miminar commented on pkg/dockerregistry/server/metrics/dispatcher.go:
```diff
@@ -0,0 +1,17 @@
+	gorillahandlers "github.com/gorilla/handlers"
+)
+
+func Dispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
```
GoDoc

miminar commented on pkg/dockerregistry/server/configuration/configuration.go:
```diff
@@ -0,0 +1,105 @@
+	Secret  string `yaml:"secret"`
+}
+
+func Parse(rd io.Reader) (*configuration.Configuration, *Configuration, error) {
```
Godoc

miminar commented on pkg/dockerregistry/server/configuration/configuration.go:
```diff
@@ -0,0 +1,105 @@
+		return nil, nil, err
+	}
+
+	return dockerConfig, &config.Openshift, nil
```
What happens if this encounters newer version than supported?

legionus replied:
@miminar I'm not sure I understand the question but I will try to answer. The docker developers  always return the [`configuration.Configuration`](https://github.com/docker/distribution/blob/master/configuration/configuration.go#L615) structure. They can convert old config to this structure, but they can't return something else.

miminar replied:
My last remark to the configuration concerns the version. We should define our own and panic if the configuration file defines any other.

miminar commented on pkg/dockerregistry/server/configuration/configuration.go:
```diff
@@ -0,0 +1,105 @@
+
+	p := configuration.NewParser("registry", []configuration.VersionedParseInfo{
+		{
+			Version: configuration.CurrentVersion,
```
Shouldn't this be different from upstream version?

legionus replied:
@miminar No. This is required to parse config file. Without it you will get [Unsupport version](https://github.com/docker/distribution/blob/master/configuration/parser.go#L125) error. Here, the entire upstream configuration is not important to us at all because we want to parse our own part of config file.

legionus replied:
@miminar But you are right and here we must specify the current version of config file.

legionus commented on pkg/dockerregistry/server/metrichandler.go:
```diff
@@ -0,0 +1,34 @@
+			},
+		}
+	}
+	extensionsRouter := app.NewRoute().PathPrefix(api.ExtensionsPrefix).Subrouter()
```
@miminar Fixed in `dockerregistry/server/api`

legionus commented on pkg/cmd/dockerregistry/dockerregistry.go:
```diff
@@ -68,33 +70,33 @@ func Execute(configFile io.Reader) {
 
 	// TODO add https scheme
-	adminRouter := app.NewRoute().PathPrefix("/admin/").Subrouter()
+	adminRouter := app.NewRoute().PathPrefix(api.AdminPrefix).Subrouter()
```
@miminar Fixed  in `dockerregistry/server/api`

miminar commented on pkg/dockerregistry/server/configuration/configuration.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+type openshiftConfig struct {
+	Openshift Configuration
```
Shouldn't this have json annotation as well?

legionus replied:
@miminar Not needed This is an internal placeholder.

miminar commented on pkg/dockerregistry/server/configuration/configuration_test.go:
```diff
@@ -0,0 +1,90 @@
+
+	if !extraConfig.Metrics.Enabled {
+		t.Fatalf("unexpected value: extraConfig.Metrics.Enabled != true")
+	}
```
Check `openshift->version` and `openshift->metrics->secret` as well.

miminar commented on pkg/dockerregistry/server/configuration/configuration.go:
```diff
@@ -0,0 +1,108 @@
+		return nil, nil, err
+	}
+
+	openshiftEnv, err := popEnv("REGISTRY_OPENSHIFT_")
```
Hmm, I wonder what to do about `"REGISTRY_OPENSHIFT_VERSION"`? :smile:

legionus replied:
@miminar good catch!

miminar commented on pkg/dockerregistry/server/configuration/configuration_test.go:
```diff
@@ -0,0 +1,90 @@
+  addr: :5000
+  relativeurls: "true"
+storage:
+  inmemory: {}
```
Let's not make the `openshift` section mandatory for now.

I know...I said no more configuration comments, but I can't resist.

legionus replied:
@miminar It's not mandatory right now. I will add test for this.

miminar replied:
Thanks!

miminar commented:
What else will be useful to measure? My tips:

- time to process single storage operation
- number of master API calls per request per resource
- time to process single API call
- number of failed API calls (not counting Unauthorized and NotFound)
- ratio of cache hits for pull-through
- ratio of successful stat of local repository link vs successful digest match against etcd
- ratio of found/not found per resource (low numer suggests something going terribly wrong)

legionus commented:
> What else will be useful to measure?

@miminar Some of these metrics are already collected. Also since the collection of metrics requires memory I would wait for the response of the administrators about what they need to see in addition. We can add many different metrics, but they can remain unclaimed by the administrators and will only be in vain to consume memory.

legionus commented:
[test]

legionus commented:
[test]

miminar commented:
@legionus most of the metrics I mentioned are the ones I'd like to see when debugging customer's problem. Admins will surely have different requirements. How much memory does it take to add 1 metric such as the one added by this PR?

legionus commented:
> How much memory does it take to add 1 metric such as the one added by this PR?

I do not have these numbers.

legionus commented:
[test]

legionus commented:
@smarterclayton since @mfojtik asked me to add this PR to the 3.6, please take a look again.

legionus commented:
[test] #13644

smarterclayton commented:
Will try and take a look tomorrow.

On Wed, Apr 5, 2017 at 2:02 PM, Alexey Gladkov <notifications@github.com>
wrote:

> [test]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/12711#issuecomment-291840079>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3ZyU_bCApFIqnbqIqBRQsHKtXUmks5rs4LfgaJpZM4LxbRp>
> .
>

mfojtik commented:
@smarterclayton my reason this should be in 3.6 is basically the ops team being able to monitor health of the registry and make scaling decisions. we don't have to make this perfect right now and we can always add more metrics to this as we go.

legionus commented:
@mfojtik @smarterclayton ping

mfojtik commented:
LGTM

Will wait for Clayton to have final word.

smarterclayton commented on pkg/dockerregistry/server/auth.go:
```diff
@@ -535,3 +545,11 @@ func verifyPruneAccess(ctx context.Context, client client.SubjectAccessReviews)
+	config := ConfigurationFrom(ctx)
+	if config.Metrics.Enabled {
+		return config.Metrics.Secret == token
+	}
```
If we're already doing auth from the registry, we should use the SAR check instead of a secret.  That allows us to then use the same policy the apiservers do.

legionus replied:
@smarterclayton In the beginning, we discussed this already and came to the decision that we would use the shared secret for metrics endpoint because router doing it in same way.

https://github.com/openshift/origin/pull/12711#issuecomment-276216300
https://github.com/openshift/origin/pull/12711#issuecomment-276679344

smarterclayton replied:
I don't agree with that.  If you are already doing auth via the master, we want metrics to use that behavior.  Static secret is only for things that aren't integrated with master auth (which is a limited set, but registry is under)

smarterclayton replied:
Actually, does the registry support cert auth (is it using the correct filter for that)?  If not, I'm ok with static secret.

legionus replied:
> If you are already doing auth via the master, we want metrics to use that behavior. Static secret is only for things that aren't integrated with master auth (which is a limited set, but registry is under)

We have already discussed [this](https://github.com/openshift/origin/pull/12711#issuecomment-276105926). Registry metrics are not part of the master API. The metrics belongs to the registry server only. You propose to make some virtual object (`registrymetrics` for example) and use SAR to check access to it ?

> Actually, does the registry support cert auth (is it using the correct filter for that)? If not, I'm ok with static secret.

The registry does not have a ready solution for it. Additionally, cert authentication will work only if HTTPS enabled. It means that it will be impossible to restrict access to metrics in an insecure registry.

@mfojtik @smarterclayton I propose to leave the possibility of shared secret authentication and add cert authentication later (in 3.7) because we don't have time to add it for 3.6. ok?

mfojtik replied:
i'm oking with it move to 3.7

smarterclayton commented on pkg/dockerregistry/server/metrics/metrics.go:
```diff
@@ -0,0 +1,50 @@
+			Namespace: registryNamespace,
+			Subsystem: registrySubsystem,
+			Name:      "request_duration_microseconds",
+			Help:      "Request latency summary in microseconds for each operation",
```
Standard Prometheus convention is for everythjng to be in seconds now.  See the Prometheus docs - there's a section that recommends naming and pattern

legionus replied:
@smarterclayton Changed it to seconds. But `http_request_duration_microseconds` will remain in place.  We have really old version of `github.com/prometheus/client_golang/prometheus` in the vendor and `InstrumentHandler` uses microseconds as unit, which is deprecated in new version and should be replaced by seconds.

smarterclayton commented:
I'm ok with shared secret for now.  Just to confirm - it can be set via
environment variable using normal Docker registry configuration override
rules?

On Wed, Apr 12, 2017 at 3:56 PM, Alexey Gladkov <notifications@github.com>
wrote:

> *@legionus* commented on this pull request.
> ------------------------------
>
> In pkg/dockerregistry/server/auth.go
> <https://github.com/openshift/origin/pull/12711#discussion_r111245288>:
>
> > @@ -535,3 +545,11 @@ func verifyPruneAccess(ctx context.Context, client client.SubjectAccessReviews)
>  	}
>  	return nil
>  }
> +
> +func isMetricsBearerToken(ctx context.Context, token string) bool {
> +	config := ConfigurationFrom(ctx)
> +	if config.Metrics.Enabled {
> +		return config.Metrics.Secret == token
> +	}
>
> If you are already doing auth via the master, we want metrics to use that
> behavior. Static secret is only for things that aren't integrated with
> master auth (which is a limited set, but registry is under)
>
> We have already discussed this
> <https://github.com/openshift/origin/pull/12711#issuecomment-276105926>.
> Registry metrics are not part of the master API. The metrics belongs to the
> registry server only. You propose to make some virtual object (
> registrymetrics for example) and use SAR to check access to it ?
>
> Actually, does the registry support cert auth (is it using the correct
> filter for that)? If not, I'm ok with static secret.
>
> The registry does not have a ready solution for it. Additionally, cert
> authentication will work only if HTTPS enabled. It means that it will be
> impossible to restrict access to metrics in an insecure registry.
>
> @mfojtik <https://github.com/mfojtik> @smarterclayton
> <https://github.com/smarterclayton> I propose to leave the possibility of
> shared secret authentication and add cert authentication later (in 3.7)
> because we don't have time to add it for 3.6. ok?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/12711#discussion_r111245288>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2elg5j-UacRy9nX0fikdWMjKbcgks5rvSx6gaJpZM4LxbRp>
> .
>

legionus commented:
> I'm ok with shared secret for now.  Just to confirm - it can be set via environment variable using normal Docker registry configuration override rules?

@smarterclayton Yes.

mfojtik commented:
@smarterclayton [merge]-ing this as we need this for perf testing. i think it will be easy to add more metrics as we go.

legionus commented:
Installer failed [merge]

legionus commented:
Ansible Task Failed [merge]

legionus commented:
flake #13271 [merge]

legionus commented:
Installer failed [merge]

legionus commented:
[merge]

legionus commented:
AWS EC2 problems [merge]

legionus commented:
[merge]

smarterclayton commented:
[merge]

On Mon, Apr 24, 2017 at 6:45 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/merge FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_origin/456/)
> (Base Commit: 53a4d1b
> <https://github.com/openshift/origin/commit/53a4d1b2f0515166eb3c6c89dabae0c92503038f>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/12711#issuecomment-296616033>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6hzjkoyoiBSLGJvboZ10TCVD289ks5rzH1DgaJpZM4LxbRp>
> .
>

mfojtik commented:
[test] in meanwhile... if we can get the test green, we can pretest-merge this...

smarterclayton commented on images/dockerregistry/config.yml:
```diff
@@ -35,3 +35,8 @@ middleware:
+  version: 1.0
+  metrics:
+    enabled: false
+    secret: TopSecretToken
```
This is incredibly unsafe and needs to be fixed :)

We should not check in default secrets - instead this should be empty, and the user MUST set both secret and enabled in order to turn it on.  That should be done by `oc adm registry`

smarterclayton replied:
It should be impossible for a user to accidentally turn this on and have the default secret in place.

mfojtik replied:
agree, this is abad practice

@legionus FYI

legionus replied:
More examples:
```
storage:
  azure:
    accountname: accountname
    accountkey: base64encodedaccountkey
  s3:
    accesskey: awsaccesskey
    secretkey: awssecretkey
  swift:
    username: username
    password: password
  oss:
    accesskeyid: accesskeyid
    accesskeysecret: accesskeysecret
health:
  http:
      headers:
        Authorization: [Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==]
proxy:
  username: username
  password: password
```
@smarterclayton @mfojtik This config file is full of plaintext passwords even if we can fix this. We have already discussed the possibility of using secrets for this file and came to the conclusion that if this is done, then the entire configuration will have to be put there.

> agree, this is abad practice

Don't tell me. In the beginning, I had not even considered this decision. You asked for a shared secret.

mfojtik replied:
@legionus the problem is the fact that 'secret: TopSecretToken' is hardcoded and looks like nobody will change it by default, so we will end up with tons of clusters with this secret :-)

legionus replied:
@mfojtik Now I see :) How about:
```
secret: Place-real-secret-here
```
?

mfojtik replied:
same problem :-) how about not having it enabled by default and update `oc registry` to allow to set this up?
